clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java

```diff
@@ -25,25 +25,40 @@
 import java.util.List;
 
 public class MetadataRequest extends AbstractRequest {
-    
+
     private static final Schema CURRENT_SCHEMA = ProtoUtils.currentRequestSchema(ApiKeys.METADATA.id);
     private static final String TOPICS_KEY_NAME = "topics";
 
     private final List<String> topics;
 
+    /**
+     * Constructor to support asking for no topic metadata
+     */
+    public MetadataRequest() {
```

Comment:
Good suggestion. I like the builder pattern as well. The implementation can be a bit verbose in java, but it can do validation at build time and prevent the telescoping constructor problem.  It may also be able to provide simplified api compatibility.